### PR TITLE
Issue 6202

### LINKED ISSUE #6202: Failure to split a line exceeding the maximum width
The following code fails to format `line formatted, but exceeded maximum width`:

```rust
impl EarlyLintPass for NeedlessContinue {
    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
        if let ExprKind::Loop(body, label, ..) | ExprKind::While(_, body, label) | ExprKind::ForLoop { body, label, .. } =
            &expr.kind
            && !in_external_macro(cx.sess, expr.span)
        {
            check_final_block_stmt(cx, body, label, expr.span.ctxt());
        }
    }
}
```

The line causing the error is the if let pattern (122 characters). Shortening the line by one character causes the same issue. Adding one character to the line causes it to be formatted correctly.

config:
```toml
max_width = 120
comment_width = 100
match_block_trailing_comma = true
wrap_comments = true
edition = "2021"
error_on_line_overflow = true
imports_granularity = "Module"
version = "Two"
ignore = ["tests/ui/crashes/ice-10912.rs"]
```

version: rustfmt 1.7.0-nightly (8337ba91 2024-06-12)

### PR DESCRIPTION
**Pull request by CalebLItalien**
This closes #6202 .

### CODE REVIEW

File: tests/source/issue-6202/issue_example.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// rustfmt-max_width: 120
+// rustfmt-comment_width: 100
+// rustfmt-match_block_trailing_comma: true
+// rustfmt-wrap_comments: true
+// rustfmt-edition: 2021
+// rustfmt-error_on_line_overflow: true
+// rustfmt-imports_granularity: Module
+// rustfmt-version: Two
```
Comment (ytmimi) [inline tests/source/issue-6202/issue_example.rs:8]: Although the user provided us with all of their configs, I think it's only the `max_width` setting that we need to reproduce the error. Could you trim this down to only include the relevant configs.

File: src/expr.rs [hunk 2]
```diff
@@ -1860,7 +1860,10 @@ fn rewrite_let(
     // TODO(ytmimi) comments could appear between `let` and the `pat`
 
     // 4 = "let ".len()
-    let pat_shape = shape.offset_left(4)?;
+    let mut pat_shape = shape.offset_left(4)?;
+    if context.config.version() == Version::Two {
+        pat_shape = pat_shape.sub_width(2)?;
```
Comment (ytmimi) [inline src/expr.rs:1865]: Let's leave a comment explaining why we're subtracting 2 from the width. Something like `// 2 for the length of " ="` should do.

**Review by ytmimi — CHANGES_REQUESTED** [review]
Let's add a comment to the code and reduce the test case down to the minimum configs needed to reproduce the issue.

File: src/expr.rs [hunk 3]
```diff
@@ -1860,7 +1860,10 @@ fn rewrite_let(
     // TODO(ytmimi) comments could appear between `let` and the `pat`
 
     // 4 = "let ".len()
-    let pat_shape = shape.offset_left(4)?;
+    let mut pat_shape = shape.offset_left(4)?;
+    if context.config.version() == Version::Two {
+        pat_shape = pat_shape.sub_width(2)?; // 2 to account for the length of " ="
```
Comment (ytmimi) [inline src/expr.rs:1865]: Nit: I'd prefer the comment to be above the line like we do in most other explanatory comments within the codebase.

**Review by ytmimi — CHANGES_REQUESTED** [review]
Just a minor nit and then I think we're good to go. Also, can you please squash all these commits and then rebase on the latest master. The team strongly discourages merge commits in PRs.

**Comment by CalebLItalien** [comment]
Apologies, was trying to squash the recent merge commit and accidentally reset the branch. I have a backup and will open a new PR.

**Comment by CalebLItalien** [comment]
Apologies, was trying to squash the recent merge commit and accidentally reset the branch. This reopens the PR.

**Comment by ytmimi** [comment]
The git history seems off to me. I think you might have squashed too many commits or grabbed the wrong commit message somehow?

**Comment by vats004** [comment]
hey can I work on this issue

**Comment by ytmimi** [comment]
@vats004 Sure! You can create a nearly identical PR. The only difference is that we'll need to gate this change on `StyleEdition::Edition2027`, instead of `Version::Two`.

**Comment by ytmimi** [comment]
> What's the rationale for gating a clear bugfix behind a future edition?

rustfmt has a strong stability guarantee

**Comment by ytmimi** [comment]
Closing since #6572 was merged instead
